demo/api/src/ml.module.ts

```diff
@@ -0,0 +1,84 @@
+import { MlRequest, Options } from "@comet/cms-api/lib/ml/ml.types";
+
+export const MLModels = () => {
+    const createMessages = (options: MlRequest, imageDetail: "low" | "high") => {
+        const messages = [];
+        options.instructions && messages.push({ role: "system", content: options.instructions });
+        options.context && messages.push({ role: "user", content: options.context });
+
+        if (options.examples) {
+            for (const example of options.examples) {
+                messages.push({
+                    role: example.role,
+                    content: example.text,
+                });
+            }
+        }
+
+        if (options.image) {
+            messages.push({
+                role: "user",
+                content: [
+                    {
+                        type: "image_url",
+                        image_url: {
+                            url: options.image.base64,
+                            detail: imageDetail,
+                        },
+                    },
+                ],
+            });
+        }
+
+        options.prompt && messages.push({ role: "user", content: options.prompt });
+
+        return messages;
+    };
+
+    function mapOptions(options?: Options) {
+        if (!options) {
+            return {};
+        }
+        return {
+            seed: options.seed,
+            temperature: options.temperature,
+            max_tokens: options.maxTokens,
+        };
+    }
+
+    const createRequest = (options: MlRequest, imageDetail: "low" | "high", apiKey: string) => {
+        const headers = new Headers();
+        headers.append("api-key", apiKey);
+        headers.append("Content-Type", "application/json");
+
+        const body = JSON.stringify({
+            messages: createMessages(options, imageDetail),
+            ...mapOptions(options.options),
+        });
+
+        return {
+            method: "POST",
+            headers: headers,
+            body: body,
+        };
+    };
+
+    const convertResponse = async (response: Response) => {
+        const content = await response.json();
+        if (content.error) {
+            throw new Error(`ML request failed - ${content.error.message}`);
+        }
+        return content.choices[0].message.content;
+    };
+
+    return {
+        image: async (options: MlRequest) => {
+            const response = await fetch("apiUrl", createRequest(options, "low", "apiKey"));
```

Comment:
It wouldn't, unless the user provides a URL and key for an OpenAI model. It is just there to show an example implementation

Reply:
I'd prefer if I could easily test/develop this feature by providing environment variables for the key and URL. We support .env.local, which can be used for this. Your module in Demo should be conditionally import if those environment variables are present.

Reply:
the demo could also have some mock implementation that simply returns a fixed string. Or fall back to that mock if API key is missing.

Reply:
Currently, I'm importing the module conditionally based on the API key and API URL. I could revert this change and always import the module but define the models conditionally. This would have the effect that if the user presses the icon and no URL and key are provided, it would always return "Model not defined" instead of a "mutation not defined" error.